File: dart/tco.dart
// Check Tail Call Optimization support

// fib: recur variant
int fib_recur(int n) {
  if (n <= 1) return 1;
  return fib_recur(n - 1) + fib_recur(n - 2);
}

// fib: tail recur variant
int fib_tail_recur(int n) {
  return visit(n, 1, 0);
}

int visit(int i, int a, int b) {
  if (i == 0) return a;
  return visit(i - 1, a + b, a);
}

// fib: iterate variant
int fib_iterate(int n) {
  var data = [1, 1];
  for (var i = 1; i < n; i++) {
    data = [data[0] + data[1], data[0]];
  }
  return data[0];
}

main(List<String> args) {
  int n = 1111;
  print('fib_iterate(${n}) = ${fib_iterate(n)}');
  print('fib_tail_recur(${n}) = ${fib_tail_recur(n)}');
  print('fib_recur(${n}) = ${fib_recur(n)}');
}

File: elixir/tco.exs
#!/usr/bin/env elixir

# check Tail Call Optimization support

defmodule Fib do
  # this would check if TCO
  defp visit(0, a, _b), do: a
  defp visit(i, a, b), do: visit(i - 1, a + b, a)

  @doc """
  tail recur variant
  """
  def tail_recur(n) when n >= 0, do: visit(n, 1, 0)

  @doc """
  recur variant
  """
  def recur(0), do: 1
  def recur(1), do: 1
  def recur(n) when n > 1, do: recur(n - 1) + recur(n - 2)
end

n = 1111
Fib.tail_recur(n) |> IO.inspect(label: "tail_recur")
Fib.recur(n) |> IO.inspect(label: "recur")
